contrib/check-for-api-changes.sh

```diff
@@ -0,0 +1,62 @@
+#!/usr/bin/env bash
+#
+# Checks the public API of crates, exits with non-zero if there are currently
+# changes to the public API not already committed to in the various api.txt
+# files.
+
+set -e
+
+export RUSTDOCFLAGS='-A rustdoc::broken-intra-doc-links'
+REPO_DIR=$(git rev-parse --show-toplevel)
+API_DIR="$REPO_DIR/api"
+
+main() {
+    check_we_have_nightly
+
+    pushd "$REPO_DIR/bitcoin" > /dev/null
+    cargo public-api --no-default-features --features=no-std > "$API_DIR/bitcoin/api-no-std.txt"
+    cargo public-api > "$API_DIR/bitcoin/api-default-features.txt"
+    cargo public-api --all-features > "$API_DIR/bitcoin/api-all-features.txt"
+    popd > /dev/null
+
+    pushd "$REPO_DIR/hashes" > /dev/null
+    cargo public-api --no-default-features > "$API_DIR/hashes/api-no-features.txt"
+    cargo public-api > "$API_DIR/hashes/api-default-features.txt"
+    cargo public-api --no-default-features --features=alloc > "$API_DIR/hashes/api-alloc.txt"
+    cargo public-api --all-features > "$API_DIR/hashes/api-all-features.txt"
+    popd > /dev/null
+
+    if is_dirty ; then
+        echo "You have introduced changes to the public API, commit the changes to api/ currently in your working directory"
+        exit 1
+    else
+        echo "No changes to the current public API"
+    fi
+}
+
+# Returns true if the api directory has been modified.
+is_dirty() {
+    if git diff --quiet api || git diff --cached api; then
+        return 1
+    else
+        return 0
+    fi
+}
+
+# cargo public-api uses nightly so the toolchain must be available.
+check_we_have_nightly() {
+    have_nightly=false
+    if cargo +nightly --version > /dev/null; then
+        have_nightly=true
+    fi
+
+    if [ "$have_nightly" = false ]; then
+        echo "script requires that a nightly toolchain to be installed (possibly >= nightly-2023-05-24)"
+        exit 1
+    fi
```

Comment:
This function could be more simply:
```
if ! cargo +nightly --version > /dev/null; then
    echo "script requires that a nightly toolchain to be installed (possibly >= nightly-2023-05-24)" >&2
    exit 1
fi
```

(Thanks ChatGPT. Other ChatGPT suggestions were "use `set -e` or `exit 1` but not both; echo error messages to stderr not stdout; don't hardcode the API path; no need to pass `$@` to main since you don't use command-line parameters; add more commens; use `cd` instead of `pushd` to be more portable. IMO none of these are worth addressing.)

Reply:
> echo error messages to stderr not stdout;

oh I like that one, I've not done that for years. Will use your suggestion, thanks.